bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp

```diff
@@ -265,8 +283,150 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
             Inst.getOpcode() == AArch64::LDRXui);
   }
 
+  bool isLDRS(const MCInst &Inst) const {
+    return (Inst.getOpcode() == AArch64::LDRSui ||
+            Inst.getOpcode() == AArch64::LDRSWui ||
+            Inst.getOpcode() == AArch64::LDRSWpre ||
+            Inst.getOpcode() == AArch64::LDRSpre ||
+            Inst.getOpcode() == AArch64::LDRSpost ||
+            Inst.getOpcode() == AArch64::LDRSWpost ||
+            Inst.getOpcode() == AArch64::LDRSWroW ||
+            Inst.getOpcode() == AArch64::LDRSWroX ||
+            Inst.getOpcode() == AArch64::LDRSroW ||
+            Inst.getOpcode() == AArch64::LDRSroX);
+  }
+
+  bool isLDRD(const MCInst &Inst) const {
+    return (Inst.getOpcode() == AArch64::LDRDui ||
+            Inst.getOpcode() == AArch64::LDRDpre ||
+            Inst.getOpcode() == AArch64::LDRDpost ||
+            Inst.getOpcode() == AArch64::LDRDroW ||
+            Inst.getOpcode() == AArch64::LDRDroX);
+  }
+
+  bool isLDRQ(const MCInst &Inst) const {
+    return (Inst.getOpcode() == AArch64::LDRQui ||
+            Inst.getOpcode() == AArch64::LDRQpre ||
+            Inst.getOpcode() == AArch64::LDRQpost ||
+            Inst.getOpcode() == AArch64::LDRQroW ||
+            Inst.getOpcode() == AArch64::LDRQroX);
+  }
+
```

Comment:
Some time ago I was already made some simmilar stuff. Please analyse this function and add misses (if any):

  bool analyzeLoad(const MCInst &Inst, MCPhysReg &From1, int64_t &Imm,
                   uint64_t &Count) const {
    // TODO add support of reg ldr, e.g. LDRXroX, LDRXroW
    From1 = AArch64::NoRegister;
    Imm = 0;
    Count = 1;
    switch(Inst.getOpcode()) {
    default:
      return false;
    case AArch64::LDRXui:
    case AArch64::LDRWui:
    case AArch64::LDRSWui:
    case AArch64::LDRHHui:
    case AArch64::LDRSHWui:
    case AArch64::LDRSHXui:
    case AArch64::LDRBBui:
    case AArch64::LDRSBWui:
    case AArch64::LDRSBXui: {
      if (!Inst.getOperand(2).isImm())
        return false;
      From1 = Inst.getOperand(1).getReg();
      Imm = Inst.getOperand(2).getImm();
      break;
    }
    case AArch64::LDURXi:
    case AArch64::LDURWi:
    case AArch64::LDURSWi:
    case AArch64::LDURHHi:
    case AArch64::LDURSHWi:
    case AArch64::LDURSHXi:
    case AArch64::LDURBBi:
    case AArch64::LDURSBWi:
    case AArch64::LDURSBXi: {
      if (!Inst.getOperand(2).isImm())
        return false;
      From1 = Inst.getOperand(1).getReg();
      Imm = Inst.getOperand(2).getImm();
      break;
    }
    case AArch64::LDRXpre:
    case AArch64::LDRWpre:
    case AArch64::LDRSWpre:
    case AArch64::LDRHHpre:
    case AArch64::LDRSHWpre:
    case AArch64::LDRSHXpre:
    case AArch64::LDRBBpre:
    case AArch64::LDRSBWpre:
    case AArch64::LDRSBXpre: {
      if (!Inst.getOperand(3).isImm())
        return false;
      From1 = Inst.getOperand(2).getReg();
      Imm = Inst.getOperand(3).getImm();
      break;
    }
    case AArch64::LDRXpost:
    case AArch64::LDRWpost:
    case AArch64::LDRSWpost:
    case AArch64::LDRHHpost:
    case AArch64::LDRSHWpost:
    case AArch64::LDRSHXpost:
    case AArch64::LDRBBpost:
    case AArch64::LDRSBWpost:
    case AArch64::LDRSBXpost: {
      From1 = Inst.getOperand(2).getReg();
      break;
    }
    case AArch64::LDPXi:
    case AArch64::LDPWi:
    case AArch64::LDPSWi: {
      if (!Inst.getOperand(3).isImm())
        return false;
      From1 = Inst.getOperand(2).getReg();
      Imm = Inst.getOperand(3).getImm();
      Count = 2;
      break;
    }
    case AArch64::LDPXpre:
    case AArch64::LDPWpre:
    case AArch64::LDPSWpre: {
      if (!Inst.getOperand(4).isImm())
        return false;
      From1 = Inst.getOperand(3).getReg();
      Imm = Inst.getOperand(4).getImm();
      Count = 2;
      break;
    }
    case AArch64::LDPXpost:
    case AArch64::LDPWpost:
    case AArch64::LDPSWpost: {
      From1 = Inst.getOperand(3).getReg();
      Count = 2;
      break;
    }
    }

    switch(Inst.getOpcode()) {
      case AArch64::LDRXui:
      case AArch64::LDPXi:
      case AArch64::LDPXpre:
        Imm <<= 3;
        break;
      case AArch64::LDRWui:
      case AArch64::LDRSWui:
      case AArch64::LDPWi:
      case AArch64::LDPSWi:
      case AArch64::LDPWpre:
      case AArch64::LDPSWpre:
        Imm <<= 2;
        break;
    }
    return true;
  }